### PR TITLE
JDK-8303861: Error handling step timeouts should never be blocked by OnError and others

### PR DESCRIPTION
**Pull request by tstuefe**
Fatal error handling is subject to several timeouts:
- a global timeout (controlled via ErrorLogTimeout)
- local error reporting step timeouts.

The latter aims to "give the JVM a kick" if it gets stuck in one particular place during error reporting. This prevents one error reporting step from hogging all the time allotted to error reporting under ErrorLogTimeout.

There are three situations where atm we suppress the global error timeout:
- if the JVM is embedded and the launcher has its abort hook installed. Obviously, that must be allowed to run.
- if the user specified one or more OnError commands to run, and these did not yet run. These must have a chance to run unmolested.
- if the user (typically developer) specified ShowMessageBoxOnError, and the error box has not yet been shown

There is a bug though, that also prevents the step timeout from firing if either condition is true. That is plain wrong.

In addition to that, the test interval WatcherThread uses to check for timeouts should be decreased. It sits at 1 second, which is too coarse-grained. 

--------

Patch:
- reworks `VMError::check_timeout()` to never block step timeouts
- adds clarifying comments
- quadruples timeout check frequency by watcher thread
- adds regression test for timeout handling with OnError
- additionally limits timeout per individual error reporting step to 5 seconds. 5 seconds is usually enough to distinguish a slow error reporting step from one that is endlessly hanging. 

Tested locally on Linux x64.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8303861](https://bugs.openjdk.org/browse/JDK-8303861): Error handling step timeouts should never be blocked by OnError and others

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [70b9add7](https://git.openjdk.org/jdk/pull/12936/files/70b9add7685b3424ac3fee55597d470aa76c8b1b)
 * [Roman Kennke](https://openjdk.org/census#rkennke) (@rkennke - **Reviewer**) ⚠️ Review applies to [70b9add7](https://git.openjdk.org/jdk/pull/12936/files/70b9add7685b3424ac3fee55597d470aa76c8b1b)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12936/head:pull/12936` \
`$ git checkout pull/12936`

Update a local copy of the PR: \
`$ git checkout pull/12936` \
`$ git pull https://git.openjdk.org/jdk pull/12936/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12936`

View PR using the GUI difftool: \
`$ git pr show -t 12936`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12936.diff">https://git.openjdk.org/jdk/pull/12936.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/runtime/nonJavaThread.cpp
Comment (dholmes-ora) [inline src/hotspot/share/runtime/nonJavaThread.cpp:274]: Harmless change but I don't see why we need sub-second resolution when the ErrorLogTimeout is in seconds. ??

**Review by dholmes-ora — APPROVED** [review]
Changes seem fine. Thanks for the clear explanation.

File: src/hotspot/share/runtime/nonJavaThread.cpp
Comment (tstuefe) [reply src/hotspot/share/runtime/nonJavaThread.cpp:274]: This matters because it also applies to each step timeout too. If we check only once per second, we overshoot each timeout by up to one second. This overshooting happens for every step timeout. If we are in a situation like here where we ignore the global step timeout, and we keep running into deadlocks e.g. in malloc, we will encounter a lot of step timeouts; these would add up.

Comment (tstuefe) [reply src/hotspot/share/runtime/nonJavaThread.cpp:274]: Note that this matters even more in the context of https://github.com/openjdk/jdk/pull/11017, since that will increase the granularity for each test, potentially exposing us to a lot more individual timeouts.

Comment (dholmes-ora) [reply src/hotspot/share/runtime/nonJavaThread.cpp:274]: I thought there was a limit of three step timeouts before we gave up? And the step timeout is a quarter of the global so still many seconds.

File: src/hotspot/share/utilities/vmError.cpp
Comment (rkennke) [inline src/hotspot/share/utilities/vmError.cpp:1787]: What is the meaning of _l here? Is it to indicate the type of the variable? If so, I would suggest to remove it. (I know it is pre-existing. I leave this up to you.)

File: test/hotspot/jtreg/runtime/ErrorHandling/TimeoutInErrorHandlingTest.java
Comment (rkennke) [inline test/hotspot/jtreg/runtime/ErrorHandling/TimeoutInErrorHandlingTest.java:40]: Isn't the better way to name this just '@test TimeoutInErrorHandlingTest' and then '@id=default' and '@id=with-on-error' ?

**Review by rkennke — APPROVED** [review]
Looks good to me. I only have 2 comments. If you change this, I don't need another review.
Thank you!
Roman

File: src/hotspot/share/utilities/vmError.cpp
Comment (tstuefe) [reply src/hotspot/share/utilities/vmError.cpp:1787]: This is a remnant of the old implementation. I'll change it.

File: test/hotspot/jtreg/runtime/ErrorHandling/TimeoutInErrorHandlingTest.java
Comment (tstuefe) [reply test/hotspot/jtreg/runtime/ErrorHandling/TimeoutInErrorHandlingTest.java:40]: Not sure what you mean? There is no @id jtreg tag. https://openjdk.org/jtreg/tag-spec.html

Comment (rkennke) [reply test/hotspot/jtreg/runtime/ErrorHandling/TimeoutInErrorHandlingTest.java:40]: > Not sure what you mean? There is no @id jtreg tag. https://openjdk.org/jtreg/tag-spec.html

Sorry, I got this wrong. The way to do it is @test id=$somestring. See for example:

https://github.com/openjdk/jdk/blob/a06426a52f16c08c95b1c0270a5fc40721921022/test/hotspot/jtreg/gc/stress/gcold/TestGCOldWithShenandoah.java#L27

Comment (tstuefe) [reply test/hotspot/jtreg/runtime/ErrorHandling/TimeoutInErrorHandlingTest.java:40]: Oh, you are right. Its also plain wrong, that specifier does nothing and gets ignored. I'll switch to "id"

File: test/hotspot/jtreg/runtime/ErrorHandling/TimeoutInErrorHandlingTest.java
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/ErrorHandling/TimeoutInErrorHandlingTest.java:51]: Do you really want the `@` in the name?

Comment (tstuefe) [reply test/hotspot/jtreg/runtime/ErrorHandling/TimeoutInErrorHandlingTest.java:51]: Argh. Typo, sorry. New keyboard.

**Comment by tstuefe** [comment]
okay this has been cooking long enough I think. Thanks @dholmes-ora and @rkennke !

/integrate

**Pull request closed without merging**
